cmd/frontend/graphqlbackend/search_results.go

```
@@ -588,19 +588,116 @@ func (r *searchResolver) evaluateLeaf(ctx context.Context) (*SearchResultsResolv
+	left.SearchResults = merged
+	// merge common search data.
+	left.searchResultsCommon.update(right.searchResultsCommon)
+	// for intersect we want the newly computed intersection size.
+	left.searchResultsCommon.resultCount = int32(len(merged))
+	return left, nil
+}
+
+func (r *searchResolver) evaluateOperator(ctx context.Context, scopeParameters []query.Node, operator query.Operator) (*SearchResultsResolver, error) {
+	if len(operator.Operands) == 0 {
+		return nil, nil
+	}
+	var new *SearchResultsResolver
+	var err error
+	result, err := r.evaluatePatternExpression(ctx, scopeParameters, operator.Operands[0])
+	if err != nil {
+		return nil, err
+	}
+	if result == nil {
+		return nil, nil
+	}
+	for _, term := range operator.Operands[1:] {
+		new, err = r.evaluatePatternExpression(ctx, scopeParameters, term)
+		if err != nil {
+			return nil, err
+		}
+		if new == nil && operator.Kind == query.And {
+			// Shortcircuit: intersecting with empty new results is empty.
+			return nil, nil
+		}
+		if new != nil {
+			switch operator.Kind {
+			case query.And:
+				result, err = intersect(result, new)
+			case query.Or:
+				result, err = union(result, new)
+			}
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+	return result, nil
+}
+
+// evaluatePatternExpression evaluates a search pattern containing and/or expressions.
+func (r *searchResolver) evaluatePatternExpression(ctx context.Context, scopeParameters []query.Node, node query.Node) (*SearchResultsResolver, error) {
+	switch term := node.(type) {
+	case query.Operator:
+		if term.Kind == query.And || term.Kind == query.Or {
+			return r.evaluateOperator(ctx, scopeParameters, term)
+		} else if term.Kind == query.Concat {
+			q := append(scopeParameters, term)
+			r.query = query.AndOrQuery{Query: q}
+			return r.evaluateLeaf(ctx)
+		}
+	case query.Parameter:
+		q := append(scopeParameters, term)
+		r.query = query.AndOrQuery{Query: q}
+		return r.evaluateLeaf(ctx)
+	}
+	// Unreachable.
+	return nil, nil
```

I don't know whether to use `panic` or not anymore.

---

If all enum types are exhausted, I think it is legitimate to panic. Especially this is internal  logic (i.e. enums are not depends on user input)

---

I agree. The trouble is that if the enum is expanded, the compiler won't warn you that this is no longer exhaustive. If we added another expression to the grammar, I am the only one working on this, I'll remember, but in general it's not guaranteed. Tests help, but also not a guarantee.

---

Or we could return an error says type not recognized, since panic can be ugly. But `return nil, nil` hides the error could have headache in debugging.

---

Yeah, I'll do that. I know I've pushed for `panic` in the past, and would still do that if it were a personal project :P But nowadays I'm so not confident we'll catch issues.